test_qmk_compiler.py

```diff
@@ -0,0 +1,337 @@
+    assert 'constuint16_tPROGMEMkeymaps' in keymap_text
+    assert 'RGB_TOG,RGB_SAI,RGB_VAI,RGB_HUD,RGB_HUI,RGB_MOD,RGB_SAD,RGB_VAD,BL_STEP' in keymap_text
+
+
+def test_0024_populate_enums_planck():
+    """Test the enum extraction code.
+    """
+    keymap_file = 'qmk_firmware/keyboards/planck/keymaps/default/keymap.c'
+    keymap_text = update_kb_redis.preprocess_source(keymap_file)
+    keymap = update_kb_redis.extract_layouts(keymap_text, keymap_file)
+    keymap_enums = update_kb_redis.populate_enums(keymap_text, keymap)
+    assert keymap_enums  ## FIXME I'm honestly not sure what update_kb_redis.populate_enums is for. Need to figure that out before I can test it.
+
+
+def test_0025_extract_layouts_cluecard():
+    """Test our layout extraction code.
+    """
+    keymap_file = 'qmk_firmware/keyboards/clueboard/card/keymaps/default/keymap.c'
+    keymap_text = update_kb_redis.preprocess_source(keymap_file)
+    layouts = update_kb_redis.extract_layouts(keymap_text, keymap_file)
+    assert layouts == 'constuint16_tPROGMEMkeymaps[][MATRIX_ROWS][MATRIX_COLS]={[0]=KEYMAP(RGB_TOG,RGB_SAI,RGB_VAI,RGB_HUD,RGB_HUI,RGB_MOD,RGB_SAD,RGB_VAD,BL_STEP,F(0),F(1),F(2))}'
+
+
+def test_0026_extract_keymap_cluecard():
+    """Test our keymap extraction code.
+    """
+    keymap_file = 'qmk_firmware/keyboards/clueboard/card/keymaps/default/keymap.c'
+    layout_macro, layer_list = update_kb_redis.extract_keymap(keymap_file)
+    assert layout_macro == 'KEYMAP'
+    assert layer_list == [['RGB_TOG', 'RGB_SAI', 'RGB_VAI', 'RGB_HUD', 'RGB_HUI', 'RGB_MOD', 'RGB_SAD', 'RGB_VAD', 'BL_STEP', 'F(0)', 'F(1)', 'F(2)']]
+
+
+def test_0027_find_layouts_cluecard():
+    """Test our layout detection code.
+    """
+    keymap_file = 'qmk_firmware/keyboards/clueboard/card/card.h'
+    layouts = update_kb_redis.find_layouts(keymap_file)
+    assert layouts == {
+        'KEYMAP': {
+            'key_count': 12,
+            'layout': [
+                {'x': 0, 'y': 0, 'w': 1, 'label': 'k00'}, {'x': 1, 'y': 0, 'w': 1, 'label': 'k01'},
+                {'x': 2, 'y': 0, 'w': 1, 'label': 'k02'}, {'x': 0, 'y': 1, 'w': 1, 'label': 'k10'},
+                {'x': 1, 'y': 1, 'w': 1, 'label': 'k12'}, {'x': 0, 'y': 2, 'w': 1, 'label': 'k20'},
+                {'x': 1, 'y': 2, 'w': 1, 'label': 'k21'}, {'x': 2, 'y': 2, 'w': 1, 'label': 'k22'},
+                {'x': 0, 'y': 3, 'w': 1, 'label': 'k11'}, {'x': 0, 'y': 4, 'w': 1, 'label': 'k30'},
+                {'x': 1, 'y': 4, 'w': 1, 'label': 'k31'}, {'x': 2, 'y': 4, 'w': 1, 'label': 'k32'}
+            ]
+        }
+    }
+
+
+def test_0028_find_info_json_clueboard_66_rev3():
+    """Make sure we can detect all the info.json files for cluecard.
+    """
+    info_json_files = update_kb_redis.find_info_json('clueboard/66/rev3')
+    assert info_json_files == ['qmk_firmware/keyboards/clueboard/66/rev3/../../info.json', 'qmk_firmware/keyboards/clueboard/66/rev3/../info.json']
```

Comment:
is there a way to run these against the equivalent of Path.Clean and remove duplicates? The relative paths make me sad.

Reply:
They make me sad too, but I'm a little loathe to add extra processing just to have prettier paths. You'd never see those paths if it weren't for the test.